ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java

```diff
@@ -219,4 +229,69 @@ private static LogicalSchema multiExpressionSchema(
     final ColumnName keyName = ColumnNames.nextKsqlColAlias(sourceSchema);
     return buildSchemaWithKeyType(sourceSchema, keyName, SqlTypes.STRING);
   }
+
+  private static final class ExpressionGrouper implements Grouper {
+
+    private final LogicalSchema schema;
+    private final ImmutableList<ExpressionMetadata> groupBys;
+    private final ProcessingLogger logger;
+
+    ExpressionGrouper(
+        final LogicalSchema sourceSchema,
+        final List<ExpressionMetadata> groupBys,
+        final ProcessingLogger logger
+    ) {
+      this.schema = expressionSchema(sourceSchema, groupBys);
+      this.groupBys = ImmutableList.copyOf(requireNonNull(groupBys, "groupBys"));
+      this.logger = Objects.requireNonNull(logger, "logger");
+
+      if (this.groupBys.isEmpty()) {
+        throw new IllegalArgumentException("Empty group by");
+      }
+    }
+
+    @Override
+    public LogicalSchema getSchema() {
+      return schema;
+    }
+
+    @Override
+    public GenericKey apply(final GenericRow row) {
+      final GenericKey.Builder builder = GenericKey.builder(groupBys.size());
+
+      for (int i = 0; i < groupBys.size(); i++) {
+        final Object result = processColumn(i, groupBys.get(i), row, logger);
+        if (result == null) {
+          return null;
+        }
+
+        builder.append(result);
+      }
+
+      return builder.build();
+    }
+  }
+
+  private static LogicalSchema expressionSchema(
```

Comment:
not your original code, but I'm wondering why we can't leverage the schema that we built in `LogicalPlanner`. It seems like these two schemas have to align perfectly or else this won't work (and the code is quite similar, this just handles `ExpressionMetadata` instead of `Expression` directly). Am I right? I haven't actually dug into it that much.

Reply:
Are you suggesting that the code for how we build the schema in the two places should be shared, or that the actual schema that gets built should be passed from `LogicalPlanner` to `GroupByParamsFactory`? I haven't looked too closely at the details of how the schema is built in `LogicalPlanner` but the latter doesn't make conceptual sense as the two are separated by the execution plan that gets serialized into the command topic.

Reply:
I did mean the latter. Doesn't the serialized execution plan contain the schema (that was generated by the logical planner) too, though?

Reply:
We don't write the full schema in, no. Here's an example historic plan: https://github.com/confluentinc/ksql/blob/db2d01963930d2fa45f26670ecd6734fbf9f79e8/ksqldb-functional-tests/src/test/resources/historical_plans/group-by_-_AVRO_primitive_key/6.2.0_1607109704822/plan.json#L109 For the group by step we only write the group by columns, which is why we have to reconstruct the rest of the schema.

Reply:
ack, out of scope then but seems like something that might be worth looking into some day